### DISCUSSION TITLE
Advice Needed: Migrating pre-existing Snowflake environment to SnowDDL management.

### DISCUSSION THREAD

**Discussion by bradleyess** [littleK0i/SnowDDL#118] [General]
Is there a clearly laid out migration path for this? I'd like to progressively bring existing databases into SnowDDL management

**Comment by littleK0i** [comment]
It highly depends on what you currently have in terms of data infrastructure.

I would normally suggest to migrate everything in one go, but make it safe & reliable by creating and testing everying in separate DEV environment.

1. Create a separate Snowflake account for testing. `CREATE ACCOUNT` is the best, but trial account can be sufficient.
2. Convert existing objects to SnowDDL YAML configs. Undocumented `snowddl-convert` command is available for DATABASE, SCHEMA, TABLE and VIEW to simplify the process a bit.
3. Think hard about access management and how your organization grants can be expressed in terms of SnowDDL schema roles / business roles / user roles. Usually it leads to massive simplification of grants.
4. Once config conversion is finished, use it to create objects in test account.
5. Compare objects in test account vs objects in prod account.
6. (Optionally) Run ETL jobs on test account using small sample of data. Make sure all ingestions and transformations work perfectly.

Once everything is done, you can now switch prod account to SnowDDL.

---

In theory, you may try to migrate slowly and grant access to objects managed outside of SnowDDL by using `global_roles`. But from my experience "slow" migrations are usually more difficult and tend to get stuck somewhere in the middle, and now you have two systems doing the same thing. 

---

If you encounter any specific problems, feel free to post questions, and we'll try to figure it out.
